### PR TITLE
feat: add traceProvider && meterProvider option

### PR DESCRIPTION
**Pull request by HoganEdwardChu**
@rueian  maybe this can be option? Thanks

### CODE REVIEW

**Comment by rueian** [comment]
Hey @HoganEdwardChu,

Thank you for your quick support for https://github.com/redis/rueidis/issues/257.

File: rueidisotel/trace.go [hunk 1]
```diff
@@ -27,12 +27,25 @@ var (
...
+		opt(o)
+	}
+
+	// Set the MeterProvider and TracerProvider if specified
+	if o.meterProvider != nil {
+		otel.SetMeterProvider(o.meterProvider)
+	}
+	if o.tracerProvider != nil {
+		otel.SetTracerProvider(o.tracerProvider)
+	}
```
Comment (rueian) [inline rueidisotel/trace.go:47]: These will set the global providers. We should not do that.
We should instead do this after applying `opts`:
```go
	if o.meterProvider == nil {
		o.meterProvider = otel.GetMeterProvider()
	}
	if o.tracerProvider == nil {
		o. tracerProvider  = otel.GetTracerProvider()
	}
```

File: rueidisotel/trace.go [hunk 2]
```diff
@@ -53,10 +66,26 @@ func TraceAttrs(attrs ...attribute.KeyValue) Option {
...
+
 type otelclient struct {
-	client rueidis.Client
-	mAttrs []attribute.KeyValue
-	tAttrs []attribute.KeyValue
+	client         rueidis.Client
+	mAttrs         []attribute.KeyValue
+	tAttrs         []attribute.KeyValue
+	meterProvider  metric.MeterProvider
+	tracerProvider trace.TracerProvider
```
Comment (rueian) [inline rueidisotel/trace.go:88]: Besides the providers, we should also move `tracer`, `cscMiss` and `cscHits` into this struct.

Comment (HoganEdwardChu) [reply rueidisotel/trace.go:88]: ah... got it. I trying to add test codes and i faced the problem what you said

File: rueidisotel/trace.go (hunk 1 above)
Comment (HoganEdwardChu) [reply rueidisotel/trace.go:52]: sorry for late you mean https://github.com/redis/rueidis/blob/1defd2dba8930c58d3f790ca89381ba370fd4111/rueidisotel/trace.go#L55 ?

File: rueidisotel/trace.go [hunk 3]
```diff
@@ -27,12 +27,26 @@ var (
...
+	}
+	if o.tracerProvider == nil {
+		o.tracerProvider = otel.GetTracerProvider() // Default to global TracerProvider
+	}
+	// Now that we have the meterProvider and tracerProvider, get the Meter and Tracer
+	o.meter = o.meterProvider.Meter(name)
+	o.tracer = o.tracerProvider.Tracer(name)
+	// Now create the counters using the meter
+	o.cscMiss, _ = o.meter.Int64Counter("rueidis_do_cache_miss")
+	o.cscHits, _ = o.meter.Int64Counter("rueidis_do_cache_hits")
```
Comment (rueian) [inline rueidisotel/trace.go:49]: So, after moving the tracer and counters to here, the original global ones should be removed.

Comment (HoganEdwardChu) [reply rueidisotel/trace.go:49]: got it

File: rueidisotel/trace.go [hunk 4]
```diff
@@ -17,93 +17,122 @@ import (
...
+func WithMetricAttrs(attrs ...attribute.KeyValue) Option {
 	return func(o *otelclient) {
 		o.mAttrs = attrs
 	}
 }
 
-// TraceAttrs set additional attributes to append to each trace.
-func TraceAttrs(attrs ...attribute.KeyValue) Option {
+// WithTraceAttrs set additional attributes to append to each trace.
+func WithTraceAttrs(attrs ...attribute.KeyValue) Option {
```
Comment (rueian) [inline rueidisotel/trace.go:59]: Oh, this is a breaking change. We can’t do that directly. We have at least to maintain an alias to the original function name.

Comment (HoganEdwardChu) [reply rueidisotel/trace.go:59]: Got it

File: rueidisotel/trace.go (hunk 3 above)
Comment (rueian) [reply rueidisotel/trace.go:44]: Hi @HoganEdwardChu,

I just looked into these tests. They all failed with nil dereferencing on the `tracer` field. So you need to add them to all `&otelclient{...}` and `&dedicated{...}`

Comment (HoganEdwardChu) [reply rueidisotel/trace.go:44]: Hi @rueian  i added it in https://github.com/redis/rueidis/blob/d29aa821a42d9a5affbe0a2217bdaa99fe715a5c/rueidisotel/trace.go#L41 and https://github.com/redis/rueidis/blob/d29aa821a42d9a5affbe0a2217bdaa99fe715a5c/rueidisotel/trace.go#L85 you mean https://github.com/redis/rueidis/blob/d29aa821a42d9a5affbe0a2217bdaa99fe715a5c/rueidisotel/trace.go#L171 here?

File: rueidisotel/trace.go [hunk 5]
```diff
@@ -2,37 +2,46 @@ package rueidisotel
 
 import (
 	"context"
+	"go.opentelemetry.io/otel/codes"
```
Comment (rueian) [inline rueidisotel/trace.go:5]: This should be in the below section

**Comment by HoganEdwardChu** [comment]
It seem's done. Now we have options for providerThanks! @rueian 
```golang
	client = WithClient(
		client,
		TraceAttrs(attribute.String("any", "label")),
		MetricAttrs(attribute.String("any", "label")),
		WithTracerProvider(tracerProvider),
		WithMeterProvider(meterProvider),
	)
```

**Comment by rueian** [comment]
Yeah! Thank you very much @HoganEdwardChu !

**Pull request merged**